maro/rl/policy/policy_manager.py

```diff
@@ -339,3 +339,193 @@ def exit(self):
+            self._proxy.send(
+                SessionMessage(
+                    MsgTag.INIT_POLICY_STATE, self._proxy.name, trainer_name,
+                    body={MsgKey.POLICY_STATE: {name: self.policy_dict[name].get_state() for name in policy_names}}
+                )
+            )
+
+    def update(self, exp_by_policy: Dict[str, ExperienceSet]):
+        exp_to_send, updated = {}, set()
+        for policy_name, exp in exp_by_policy.items():
+            self._num_experiences_by_policy[policy_name] += exp.size
+            self._exp_cache[policy_name].extend(exp)
+            if (
+                self._exp_cache[policy_name].size >= self.update_trigger[policy_name] and
+                self._num_experiences_by_policy[policy_name] >= self.warmup[policy_name]
+            ):
+                exp_to_send[policy_name] = self._exp_cache.pop(policy_name)
+                updated.add(policy_name)
+
+        # 1. prepare exp data for each trainer node
+        msg_body_by_dest = defaultdict(dict)
+        for policy_name, exp in exp_to_send.items():
+            trainer_id_list = self._policy2trainer[policy_name]
+            if len(trainer_id_list) == 1:  # single node
+                trainer_id = trainer_id_list[0]
+                if MsgKey.EXPERIENCES not in msg_body_by_dest[trainer_id]:
+                    msg_body_by_dest[trainer_id][MsgKey.EXPERIENCES] = {}
+                msg_body_by_dest[trainer_id][MsgKey.EXPERIENCES][policy_name] = exp
+                self._logger.info(f'policy {policy_name}, exp.size = {exp.size}')
+            else:
+                for i, trainer_id in enumerate(trainer_id_list):
+                    if MsgKey.EXPERIENCES not in msg_body_by_dest[trainer_id]:
+                        msg_body_by_dest[trainer_id][MsgKey.EXPERIENCES] = {}
+                    sub_exp = exp[i::len(trainer_id_list)]
+                    msg_body_by_dest[trainer_id][MsgKey.EXPERIENCES][policy_name] = sub_exp
+                    self._logger.info(f'policy {policy_name}, sub_exp.size = {sub_exp.size}')
+
+        # 2. scatter data and receive reply of each trainer node
+        trackers = []
+        for reply in self._proxy.scatter(MsgTag.SCATTER_EXP, SessionType.TASK, list(msg_body_by_dest.items())):
+            trackers.append(reply.body[MsgKey.TRACKER])
+
+        # get gradient and update
+        # TODO: handle various config.train_epochs among policies.
+        name = list(self.policy_dict.keys())[0]
+        for _ in range(self.policy_dict[name].config.train_epochs):
+            msg_body_loss = defaultdict(dict)
+            for reply in self._proxy.scatter(MsgTag.GET_GRAD, SessionType.TASK, list(msg_body_by_dest.items())):
+                # 3. aggregate gradient
+                for policy_name, grad_dict in reply.body[MsgKey.GRAD].items():
+                    trainer_id_list = self._policy2trainer[policy_name]
+                    if len(trainer_id_list) == 1:  # single node
```

Comment:
could the code for single node and multiple nodes be unified?
maybe the operation of `/ len(trainer_id_list)` could be operated once at the end?